worldedit-core/src/main/java/com/sk89q/worldedit/util/translation/TranslationManager.java

```diff
@@ -0,0 +1,160 @@
+ * under the terms of the GNU Lesser General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.sk89q.worldedit.util.translation;
+
+import com.google.common.io.Files;
+import com.google.common.io.Resources;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.reflect.TypeToken;
+import com.sk89q.worldedit.WorldEdit;
+import com.sk89q.worldedit.util.formatting.text.Component;
+import com.sk89q.worldedit.util.formatting.text.renderer.FriendlyComponentRenderer;
+import com.sk89q.worldedit.util.io.ResourceLoader;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.text.MessageFormat;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Handles translations for the plugin.
+ *
+ * These should be in the following format:
+ * plugin.component.message[.meta]*
+ *
+ * Where,
+ * plugin = worldedit
+ * component = The part of the plugin, eg expand
+ * message = A descriptor for which message, eg, expanded
+ * meta = Any extra information such as plural/singular (Can have none to infinite)
+ */
+public class TranslationManager {
+
+    private static final Gson gson = new GsonBuilder().create();
+    private static final Type STRING_MAP_TYPE = new TypeToken<Map<String, String>>() {}.getType();
+
+    private final Map<Locale, Map<String, String>> translationMap = new HashMap<>();
+    private final FriendlyComponentRenderer<Locale> friendlyComponentRenderer = FriendlyComponentRenderer.from(
+            (locale, key) -> new MessageFormat(getTranslationMap(locale).getOrDefault(key, key), locale));
+    private Locale defaultLocale = Locale.ENGLISH;
+
+    private final WorldEdit worldEdit;
+
+    private final Set<Locale> checkedLocales = new HashSet<>();
+
+    public TranslationManager(WorldEdit worldEdit) {
+        this.worldEdit = worldEdit;
+    }
+
+    public void setDefaultLocale(Locale defaultLocale) {
+        this.defaultLocale = defaultLocale;
+    }
+
+    private Map<String, String> filterTranslations(Map<String, String> translations) {
+        translations.entrySet().removeIf(entry -> entry.getValue().isEmpty());
+        return translations;
+    }
+
+    private Map<String, String> parseTranslationFile(File file) throws IOException {
```

Comment:
Probably better to merge these into one `parseTranslationFile(InputStream)`, use `InputStreamReader` and pass that to GSON.